File: src/main/java/com/trip_service/trip_service/users/controllers/UserController.java
package com.trip_service.trip_service.users.controllers;

import com.trip_service.trip_service.utils.DTO.SendResponse;
import com.trip_service.trip_service.users.DTO.RequestBodyDTO;
import com.trip_service.trip_service.users.models.Users;
import com.trip_service.trip_service.users.services.UserServices;
import com.trip_service.trip_service.utils.exceptions.GeneratedApiException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/users")
public class UserController {

    @Autowired
    private UserServices userServices;

    @PostMapping("/create")
    public SendResponse<Users> createUser(@RequestBody Users user) {

        try {
            Users createdUser = userServices.createUser(user);

            return new SendResponse<>(200, "User Created Successfully!", "", createdUser);

        }
        catch (GeneratedApiException apiException){
            return new SendResponse<>(apiException.getStatusCode(), apiException.getMessage(), apiException.getErrorMsg(), null);
        }
        catch (Exception e) {
            return new SendResponse<>(500, e.getMessage(),"", null);
        }
    }

    @PostMapping("/login")
    public SendResponse<String> loginUser(@RequestBody RequestBodyDTO.GetLoginUserRequestBody requestBody ) {

        try {

            String response = userServices.loginUser(requestBody.getUsername(), requestBody.getPassword());

            if(response != null )  {
                return new SendResponse<>(200, "Login Successful!", "", response);
            }

            return  new SendResponse<>(204, "Login Failed!!", "", null);

        }
        catch (GeneratedApiException apiException){
            return new SendResponse<>(apiException.getStatusCode(), apiException.getMessage(), apiException.getErrorMsg(), null);
        }
        catch (Exception e) {
            return new SendResponse<>(500, e.getMessage(), "", null);
        }
    }

    @PostMapping("/list")
    @PreAuthorize("hasAnyAuthority('ADMIN')")
    public SendResponse<List<Users>> getAllUsers() {
        return new SendResponse<>(200, "Data Retrived Successfully!", "", userServices.getAllUser());
    }

    @PostMapping("/get-user-info")
    public SendResponse<Optional<Users>> getUserById(@RequestBody RequestBodyDTO.GetUserByIdRequestBody requestBody) {

        try {
            Optional<Users> foundUser = userServices.getUserById(requestBody.getUsername());

            return new SendResponse<>(200, "User Details Found!!", "", foundUser);

        }
        catch (GeneratedApiException apiException){
            return new SendResponse<>(apiException.getStatusCode(), apiException.getMessage(), apiException.getErrorMsg(), null);
        }
        catch (Exception e) {
            return new SendResponse<>(500, e.getMessage(), "", null);
        }

    }

    @PostMapping("/manipulate-user-role")
    @PreAuthorize("hasAnyAuthority('ADMIN')")
    public SendResponse<List<String>> manipulateRole( @RequestBody RequestBodyDTO.RoleRequestBody requestBody ) {

        try {

            // TODO: Sanities changeType from requestBody


            return new SendResponse<>(
                    200,
                    "Role Manipulated Successfully!!",
                    "",
                    userServices.manipulateRole( requestBody.getChangeType(), requestBody.getUsername(), requestBody.getRole())
            );

        }
        catch (GeneratedApiException apiException){
            return new SendResponse<>(apiException.getStatusCode(), apiException.getMessage(), apiException.getErrorMsg(), null);
        }
        catch (Exception e) {
            return new SendResponse<>(500, "Internal Server Error!!", e.getMessage(), null);
        }

    }

}

File: src/main/kotlin/com/trip_service/trip_service/core/users/controllers/UserController.kt
package com.trip_service.trip_service.core.users.controllers


import com.trip_service.trip_service.core.users.models.Users
import com.trip_service.trip_service.core.users.dto.RequestBodyDTO
import com.trip_service.trip_service.core.users.dto.UsersResponseBody
import com.trip_service.trip_service.core.users.services.UserService
import com.trip_service.trip_service.helpers.dto.GenerateResponse
import com.trip_service.trip_service.helpers.errors.GenerateApiException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/users")
class UserController {

    @Autowired
    private lateinit var userService: UserService;

    @GetMapping
    fun default(): String {
        return "User Service is Up and Running"
    }

    @PostMapping("/create")
    fun createUser(@RequestBody user: Users):  GenerateResponse<String> {

        try {
            return GenerateResponse( 200,userService.createUser(user), "")
        }
        catch(exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage
            )
        }

    }

    @PostMapping("/login")
    fun login(@RequestBody loginUserReqBody: RequestBodyDTO.LoginUser): GenerateResponse<String>{

        try {

            return GenerateResponse(
                200,
                "Login Successful!",
                "",
                userService.login(loginUserReqBody.username, loginUserReqBody.password)
            )

        }
        catch (exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage
            )
        }

    }

    @GetMapping("/user-info/{username}")
    @PreAuthorize("hasAnyAuthority('ADMIN')")
    fun getUserInfo(@PathVariable username: String): GenerateResponse<UsersResponseBody.UserInfo> {

        try {

            val foundUser: Users = userService.getUserById(username);

            return GenerateResponse(
                200, "User Details Found!", "", foundUser.getFilteredData()
            )

        }
        catch (exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage
            )
        }

    }

    @GetMapping("/all")
    fun getAllUser(): GenerateResponse<List<UsersResponseBody.UserInfo>> {
        return GenerateResponse(
            200,
            "Data Retrived Successfully!",
            "",
            userService.getAllUsers()
        )
    }

    @PostMapping("/manipulate-user-role")
    fun manipulateUserRole(@RequestBody manipulateUserRolesReqBody: RequestBodyDTO.ManipulateRole): GenerateResponse<Set<String>> {
        try {
            return GenerateResponse(200, "Role Manipulated Successfully!", "", userService.manipulateRole(
                manipulateUserRolesReqBody.username,
                manipulateUserRolesReqBody.role,
                manipulateUserRolesReqBody.changeType
            ))
        }
        catch (exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage
            )
        }
    }

}